File: 2sat/2sat.cpp
/* Modex ACM c.durr (2008)
   
   resoudre une formule 2-SAT en temps linéaire.

   Formule(n) crée une formule sur n variables booléens x1,x2,...,xn.
   Les litéraux sont codés 1,..,n ou les variables positives 
   et -1,..,-n pour leurs négations.
   add(x,y) ajoute la clause x OU y, où x,y sont des litéraux.
   solve() résoud la forrmule. 
   et get permet alors de lire la valeur des variables dans la solution.
*/
#include <iostream>
#include <fstream>
#include <vector>
#include <stack>
#include <assert.h>

using namespace std;


// ---------------- le graphe oriente sur les literaux

struct Sommet {
  vector<int> out;
  vector<int> in;
  int         comp; // numero de la composante fortement connexe
  int         name; // nom pour affichage
};

struct Graphe : vector<Sommet> {

  Graphe(int n) :vector<Sommet>(n) {}

  // ajouter un arc (u,v)
  void add(int u, int v) {
    (*this)[u].out.push_back(v);
    (*this)[v].in.push_back(u);
  }

};
  
// ------------------ Formule


struct Formule {
  int n; // nb variables, les variables sont de 1 a n, leur negations de -1 a -n
  Graphe G; // graphe des implications

  vector<int> var; // valeur des variables, indices de 1 a n

  enum ValPossibles {FAUX=-1, UNDEF=0, VRAI=+1};


  // given a literal, returns the vertex number
  int vertex(int x) {
    assert(1<=abs(x) && abs(x)<=n);
    return x>0 ? x-1 : 2*n+x;
  }

  // given a literal, returns its value
  int get(int x) const {
    assert(1<=abs(x) && abs(x)<=n);
    return x>0 ? var[x] : -var[-x];
  }

  // set the value of a literal
  // dis-donc, pourquoi je me met a commenter en anglais ?
  void set(int x, int val) {
    assert(1<=abs(x) && abs(x)<=n);
    if (x>0)
      var[x]  =  val;
    else
      var[-x] = -val;
  }

  Formule (int _n) :G(2*_n) , var(_n+1, UNDEF){
    n=_n;
    for (int x=1; x<=n; x++) {
      G[vertex( x)].name =  x;
      G[vertex(-x)].name = -x;
    }
  }

  // ajoute la clause x || y, qui est equivalente a -x=>y et -y=>x
  void add(int x, int y) {
    assert(1<=abs(x) && abs(x)<=n);
    assert(1<=abs(y) && abs(y)<=n);
    G.add(vertex(-x),vertex( y));
    G.add(vertex(-y),vertex( x));
  }
  
#define PAS_ENCORE -1

  stack<int>  aTraiter;
  int         heure;
  
  /* V(v).comp contient soit PAS_ENCORE si sommet pas encore visite
     sinon la date de debut de traitement par dfs (decale par -size()-1).
     V(v).comp contient sinon le numero d'un sommet representatif de la
     composante.
  */
  int sccpFrom(int u) {
    int v;
    aTraiter.push(u);
    int m = G[u].comp = heure++;
    for (int i=0; i<G[u].out.size(); i++) {
      int v = G[u].out[i];
      int mv = (G[v].comp==PAS_ENCORE) ? sccpFrom(v) : G[v].comp;
      m = min(m, mv);
    }
    if (m == G[u].comp) {
      int uval = get(G[u].name);
      int compval = (uval==UNDEF) ? VRAI : uval;
      do {
	v = aTraiter.top();
	aTraiter.pop();
	G[v].comp = u;
	set(G[v].name, compval);
      } while (u != v);
    }
    return m;
  }

  // retourne true si la formule est satisfiable
  bool solve() {
    heure = -1-G.size();
    var.assign(var.size(), UNDEF);
    for (int u=0; u<G.size(); u++)
      G[u].comp = PAS_ENCORE;

    for (int u=0; u<G.size(); u++)
      if (G[u].comp==PAS_ENCORE)
	sccpFrom(u);

    // est-ce qu'il existe une variable x et -x qui sont dans la 
    // meme componsante ?
    for (int x=1; x<=n; x++) 
      if (G[vertex(x)].comp == G[vertex(-x)].comp)
	return false;

    print("tmp.dot");
    return true;
  }


  /** afficher en format DOT le graphe d'implication avec les composantes
      fortement connexes et en indiquant les valeurs booleennes de celles-ci
  */
  void print(const char *filename) const {
    ofstream o(filename, ios::out);
    vector<int>  c[G.size()];
    o << "digraph {" << endl
      << "  node [shape=circle]" << endl;
    // produire une liste de sommets par composante
    for (int u=0; u<G.size(); u++) 
      c[G[u].comp].push_back(u); 

    // afficher les composantes
    for (int j=0; j<G.size(); j++) 
      if (c[j].size() != 0) {
	o << "  subgraph cluster" << j << " {\n    ";
	if (get(G[c[j][0]].name)==VRAI)
	  o << "style=filled; color=lightgrey;\n    ";
	for (int i=0; i<c[j].size(); i++) {
	  int u = c[j][i];
	  o << u << " [label=" << G[u].name << "]; ";
	}
	o << endl << "  }" << endl;
      }
    // afficher les arcs
    for (int u=0; u<G.size(); u++) 
      for (int i=0; i<G[u].out.size(); i++) {
	int v = G[u].out[i];
	o << "  " << u << " -> " << v << endl;
      }
    o << "}" << endl;
  }
};

// ------------------ Programme principal
int main() {
  int n, x,y;
  cin >> n;
  Formule F(n);
  while (cin >> x >> y) 
    F.add(x, y);
  if (!F.solve())
    cout << "Formule non satisfiable";
  else {
    cout << "Formule satisfiable" << endl;
    for (int x=1; x<=n; x++)
      cout << "x"<<x<< "=" << (F.get(x)==Formule::VRAI) << " ";
  }
  cout << endl;
}

File: 2sat/2sat.java
/* Modex ACM c.durr (2008)
   
   resoudre une formule booleenne 2-SAT par Tarjan en temps lineaire.
   Mais cette implementation utilise une matrice d'adjacence est sera
   donc en temps quadratique (en nombre de variables).

   format de l'entree:
        chaque instance est code par
        <n> <m> 
        m fois <x> <y>
        et la fin est marquee par
        0 0
        Dans l'instance il y a n variables, et m clauses
        x,y sont des literaux. 1..n sont les literaux positifs 
        et -1..-n les literaux negatifs.
*/

import java.util.*;

class Main {    
//snip{ snipSccp
    static int n; // nb literaux, variable xi=2*i, literal non(xi)=2*i+1
    //        pour un literal y, non(y)=y^1    
    static boolean adj[][];    // adj[x][y] = il existe un arc x=>y

    static int o[];            // o[i] = i-eme sommet traite par dfs
    static boolean deja[];     // marqueur pour dfs
    static int time;
    static int sccp[];         // sccp[u]=numero composante fort. connex. de u

    // parcours dfs normal, noter dans o l'ordre fin visite
    static void dfs(int u) {
        deja[u] = true;
        for (int v=0; v<n; v++) 
            if (adj[u][v] && !deja[v])
                dfs(v);
        o[time++] = u;    
    }

    // parcours dfs en inversant les arcs, mettre tout le monde dans meme comp.
    static void dfsInv(int u, int comp) {
        deja[u] = true;
        sccp[u] = comp;
        for (int v=0; v<n; v++) 
            if (adj[v][u] && !deja[v])
                dfsInv(v, comp);
    }

    // calculer les composantes fortement connexes, et retourner leur nombre
    static int sccp() {
        time = 0;
        deja = new boolean[n];
        o    = new int[n];
        for (int u=0; u<n; u++)
            if (!deja[u])
                dfs(u);

        int comp=0;
        deja = new boolean[n];
        sccp = new int[n];
        for (int i=n-1; i>=0; i--) {
            int u = o[i];
            if (!deja[u])
                dfsInv(u, comp++);
        }
        return comp;
    }
    //snip}

    //snip{ snip2sat
    static int nc;           // nb composantes fortement connexes
    static int val[];        // valeur d'une comp. +1=vrai, -1=faux, 0=indef.
    static int neg[];        // neg[j]=composante opposee de j

    // retourne true si la formule est satisfiable
    // dans ce cas, pour chque literal x sa valeur est val[sccp[x]]
    static boolean satisfiable() {
        nc  = sccp();
        neg = new int[nc];

        // est-ce qu'il existe une variable x et sa negation non(x)
        // qui soient dans la meme componsante ?
        for (int x=0; x<n; x+=2) 
            if (sccp[x] == sccp[x^1])
                return false;      
            else {
                neg[sccp[x]] = sccp[x^1];
                neg[sccp[x^1]] = sccp[x];
            }

        // on utilise le fait que les compsantes portent un numero dans 
        // l'ordre topologique inverse
        val = new int[nc];
        for (int j=0; j<nc; j++)
            if (val[j]==0) {
                val[j]      = -1;
                val[neg[j]] = +1;
            }
        return true;
    }
    //snip}

    public static void main(String[] args) {
        int m, x,y;
        Scanner in = new Scanner(System.in);
        for (int t=1; true; t++) {      
            n = 2*in.nextInt(); 
            m = in.nextInt();
            if (n==0)
                return;
            adj = new boolean [n][n];
            while (m-->0) {
                x = in.nextInt();
                x = (x>0) ? 2*(x-1) : 1-2*(x+1);
                y = in.nextInt();
                y = (y>0) ? 2*(y-1) : 1-2*(y+1);
                adj[x^1][y] = adj[y^1][x] = true;
            }
            if (!satisfiable())
                System.out.println("Formule "+t+" non satisfiable");
            else {
                System.out.println("Formule "+t+" satisfiable");
                for (x=0; x<n; x+=2) 
                    System.out.println("  X"+(x/2+1)+"="+(val[sccp[x]]==+1));
            }
        }
    }
}
